File: python/tests/test_hamiltonian.py
import unittest

from snake_game.ai.hamiltonian import build_hamiltonian_cycle


class HamiltonianCycleTests(unittest.TestCase):
    def test_visits_every_cell_exactly_once(self):
        cols, rows = 6, 4
        cycle = build_hamiltonian_cycle(cols, rows)
        self.assertEqual(len(cycle), cols * rows)
        self.assertEqual(len(set(cycle)), cols * rows)

    def test_consecutive_cells_are_grid_adjacent_including_the_wraparound(self):
        cols, rows = 6, 4
        cycle = build_hamiltonian_cycle(cols, rows)
        for a, b in zip(cycle, cycle[1:] + cycle[:1]):
            manhattan_distance = abs(a[0] - b[0]) + abs(a[1] - b[1])
            self.assertEqual(manhattan_distance, 1)

    def test_matches_the_actual_game_board_dimensions(self):
        cycle = build_hamiltonian_cycle(18, 12)
        self.assertEqual(len(cycle), 18 * 12)

    def test_rejects_a_grid_with_an_odd_number_of_cells(self):
        with self.assertRaises(ValueError):
            build_hamiltonian_cycle(3, 3)


if __name__ == '__main__':
    unittest.main()

File: cpp/tests/test_hamiltonian.cpp
#include <algorithm>
#include <cstdlib>
#include <stdexcept>
#include <unordered_set>

#include "ai/hamiltonian.hpp"
#include "doctest.h"

TEST_CASE("visits every cell exactly once") {
    int cols = 6, rows = 4;
    auto cycle = hamiltonian::BuildCycle(cols, rows);
    CHECK(cycle.size() == static_cast<std::size_t>(cols * rows));

    std::unordered_set<Cell, CellHash> unique(cycle.begin(), cycle.end());
    CHECK(unique.size() == static_cast<std::size_t>(cols * rows));
}

TEST_CASE("consecutive cells are grid-adjacent, including the wraparound") {
    auto cycle = hamiltonian::BuildCycle(6, 4);
    for (std::size_t i = 0; i < cycle.size(); ++i) {
        const Cell& a = cycle[i];
        const Cell& b = cycle[(i + 1) % cycle.size()];
        int manhattan = std::abs(a.col - b.col) + std::abs(a.row - b.row);
        CHECK(manhattan == 1);
    }
}

TEST_CASE("matches the actual game board dimensions") {
    auto cycle = hamiltonian::BuildCycle(18, 12);
    CHECK(cycle.size() == 18u * 12u);
}

TEST_CASE("rejects a grid with an odd number of cells") {
    CHECK_THROWS_AS(hamiltonian::BuildCycle(3, 3), std::invalid_argument);
}
